core/bounded_buffer.go

```diff
@@ -3,45 +3,55 @@
 
 package core
 
-import (
-	"github.com/ethereum/go-ethereum/common"
-)
-
-// BoundedBuffer keeps [size] common.Hash entries in a buffer and calls
-// [callback] on any item that is evicted. This is typically used for
+// BoundedBuffer keeps [size] entries of type [K] in a buffer and calls
+// [callback] on any item that is overwritten. This is typically used for
 // dereferencing old roots during block processing.
-type BoundedBuffer struct {
+//
+// BoundedBuffer is not thread-safe and requires the caller synchronize usage.
+type BoundedBuffer[K any] struct {
 	lastPos  int
 	size     int
-	callback func(common.Hash)
-	buffer   []common.Hash
+	callback func(K)
+	buffer   []K
+
+	cycled bool
 }
 
 // NewBoundedBuffer creates a new [BoundedBuffer].
-func NewBoundedBuffer(size int, callback func(common.Hash)) *BoundedBuffer {
-	return &BoundedBuffer{
+func NewBoundedBuffer[K any](size int, callback func(K)) *BoundedBuffer[K] {
+	return &BoundedBuffer[K]{
+		lastPos:  -1,
 		size:     size,
 		callback: callback,
-		buffer:   make([]common.Hash, size),
+		buffer:   make([]K, size),
 	}
 }
 
-// Insert adds a new common.Hash to the buffer. If the buffer is full, the
-// oldest common.Hash will be evicted and [callback] will be invoked.
-//
-// WARNING: BoundedBuffer does not support the insertion of empty common.Hash.
-// Inserting such data will cause unintended behavior.
-func (b *BoundedBuffer) Insert(h common.Hash) {
-	nextPos := (b.lastPos + 1) % b.size // the first item added to the buffer will be at position 1
```

Comment:
Removed this weird off-by-1 start and now explicitly return in `Last` if the value was set